src/jit/jitstd/algorithm.h

```diff
@@ -46,4 +46,191 @@ Function for_each(InputIterator first, InputIterator last, Function func)
+            //
+            // It's not possible for newFirst to go past the end of the sort range:
+            //   - If newFirst reaches the pivot before newLast then the pivot is
+            //	   swapped to the right and we'll stop again when we reach it.
+            //   - If newLast reaches the pivot before newFirst then the pivot is
+            //	   swapped to the left and the value at newFirst will take its place
+            //     to the right so less(newFirst, pivot) will again be false when the
+            //     old pivot's position is reached.
+            do
+            {
+                ++newFirst;
+            } while ((newFirst != pivot) && less(*newFirst, *pivot));
+
+            // Find newLast such that *newLast <= *pivot.
+            //
+            // Like above, this stops when the pivot is reached and also does not
+            // go before the start of the sort range.
+            do
+            {
+                --newLast;
+            } while ((newLast != pivot) && less(*pivot, *newLast));
+
+            // If newFirst reaches newLast then we're done.
+            if (newFirst >= newLast)
+            {
+                break;
+            }
+
+            // We now have *newLast <= *pivot <= *newFirst so we need to swap
+            // *newFirst and *newLast.
+            swap(*newFirst, *newLast);
+
+            // pivot is an iterator and not the actual value, if the value gets
+            // swapped we need to update the iterator to point to the new place.
+            if (pivot == newFirst)
+            {
+                pivot = newLast;
+            }
+            else if (pivot == newLast)
+            {
+                pivot = newFirst;
+            }
+        }
+
+        RandomAccessIterator leftFirst = first;
+        RandomAccessIterator leftLast = newLast;
+        RandomAccessIterator rightFirst = newLast + 1;
+        RandomAccessIterator rightLast = last;
+
+        assert(depth < _countof(firstStack));
+
+        // Ideally, the 2 partitions should have the same size, that would guarantee
+        // log2(n) stack space. If that's not the case then push the larger partition
+        // onto the stack and sort the smaller one using tail recursion.
+        if ((leftLast - leftFirst) < (rightLast - rightFirst))
+        {
+            firstStack[depth] = rightFirst;
+            lastStack[depth] = rightLast;
+
+            first = leftFirst;
+            last = leftLast;
+        }
+        else
+        {
+            firstStack[depth] = leftFirst;
+            lastStack[depth] = leftLast;
+
+            first = rightFirst;
+            last = rightLast;
+        }
+
+        depth++;
+    }
+}
+}
+
+template<typename RandomAccessIterator, typename Less>
+void sort(RandomAccessIterator first, RandomAccessIterator last, Less less)
+{
+    assert(first <= last);
+    assert((last - first) < INT32_MAX);
+
+    if (first != last)
+    {
+        quick_sort(first, last - 1, less);
```

Comment:
Am I right that the arguments for `sort` are [first, last), but for `quick_sort` they are [first, last]?

Could you please add headers for them about it?

Reply:
Will do. Somehow I always found [first, last] preferrable for sorting, perhaps due to things like pivot selection.